src/gen/conditions.rs

```diff
@@ -0,0 +1,1568 @@
+                if a.atom(id) != a.atom(parent_id) {
+                    return Err(ValidationErr(c, ErrorCode::AssertMyParentIdFailed));
+                }
+            }
+            Condition::AssertMyPuzzlehash(hash) => {
+                if a.atom(hash) != a.atom(puzzle_hash) {
+                    return Err(ValidationErr(c, ErrorCode::AssertMyPuzzlehashFailed));
+                }
+            }
+            Condition::CreateCoinAnnouncement(msg) => {
+                ann.coin.insert((npc.coin_id.clone(), msg));
+            }
+            Condition::CreatePuzzleAnnouncement(msg) => {
+                ann.puzzle.insert((npc.puzzle_hash, msg));
+            }
+            Condition::AssertCoinAnnouncement(msg) => {
+                ann.assert_coin.insert(msg);
+            }
+            Condition::AssertPuzzleAnnouncement(msg) => {
+                ann.assert_puzzle.insert(msg);
+            }
+            Condition::AggSigMe(_, _) | Condition::AggSigUnsafe(_, _) => {
+                npc.agg_sigs.insert(cva);
+            }
+        }
+    }
+
+    Ok(npc)
+}
+
+// This function parses, and validates aspects of, the above structure and
+// returns a list of all spends, along with all conditions, organized by
+// condition op-code
+pub fn parse_spends(
+    a: &Allocator,
+    spends: NodePtr,
+    mut max_cost: Cost,
+    flags: u32,
+) -> Result<Vec<Npc>, ValidationErr> {
+    let mut ret = Vec::<Npc>::new();
+
+    // this object tracks which ranges of the heap we've scanned for zeros, to
+    // avoid scanning the same ranges multiple times.
+    let mut range_cache = RangeSet::new();
+
+    // this is where we collect all coin/puzzle announces (both create and
+    // asserts)
+    let mut ann = AnnounceState::new();
+
+    // all coin IDs that have been spent so far. When we parse a spend we also
+    // compute the coin ID, and stick it in this set. It's reference counted
+    // since it may also be referenced by announcements
+    let mut spent_coins = HashSet::<Arc<[u8; 32]>>::new();
+
+    let (mut iter, _) = next(a, spends)?;
+    while let Some((spend, next)) = pair(a, iter) {
+        iter = next;
+        // max_cost is passed in as a mutable reference and decremented by the
+        // cost of the condition (if it has a cost). This let us fail as early
+        // as possible if cost is exceeded
+        ret.push(parse_npc(
+            a,
+            &mut ann,
+            &mut spent_coins,
+            spend,
+            flags,
+            &mut max_cost,
+            &mut range_cache,
+        )?);
+    }
+
+    // check all the assert announcements
+    // if there are no asserts, there is no need to hash all the announcements
+    if !ann.assert_coin.is_empty() {
+        let mut announcements = HashSet::<[u8; 32]>::new();
+
+        // TODO:
+        // in order to avoid hashing the coin ID multiple times, do it once and
+        // then copy the hasher state to add the message for each announcement
```

Comment:
Nothing happens until 512 bits are passed to `.update`, so this optimization wouldn't do very much. See also https://github.com/in3rsha/sha256-animation

Reply:
I removed these comments